source/franz.go

```diff
@@ -12,51 +12,79 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//go:generate mockgen -destination franz_mock.go -package source -mock_names=Client=MockClient . Client
+
 package source
 
 import (
 	"context"
 	"errors"
 	"fmt"
 	"sync"
+	"time"
 
 	sdk "github.com/conduitio/conduit-connector-sdk"
 	"github.com/twmb/franz-go/pkg/kgo"
 )
 
 type FranzConsumer struct {
-	client *kgo.Client
+	client Client
 	acker  *batchAcker
 
 	iter *kgo.FetchesRecordIter
 }
 
+// Client is a franz-go kafka client.
+type Client interface {
+	Close()
+	CommitRecords(ctx context.Context, rs ...*kgo.Record) error
+	OptValue(opt any) any
+	PollFetches(ctx context.Context) kgo.Fetches
+	AllowRebalance()
+}
+
 var _ Consumer = (*FranzConsumer)(nil)
 
 func NewFranzConsumer(ctx context.Context, cfg Config) (*FranzConsumer, error) {
+	c := &FranzConsumer{
+		iter: &kgo.FetchesRecordIter{}, // empty iterator is done
+	}
+
 	opts := cfg.FranzClientOpts(sdk.Logger(ctx))
 	opts = append(opts, []kgo.Opt{
 		kgo.ConsumerGroup(cfg.GroupID),
 		kgo.ConsumeTopics(cfg.Topics...),
+		kgo.BlockRebalanceOnPoll(),
 	}...)
 
 	if !cfg.ReadFromBeginning {
 		opts = append(opts, kgo.ConsumeResetOffset(kgo.NewOffset().AtEnd()))
 	}
 	if cfg.GroupID != "" {
-		opts = append(opts, kgo.DisableAutoCommit()) // TODO research if we need to add OnPartitionsRevoked (see DisableAutoCommit doc)
+		opts = append(opts, []kgo.Opt{
+			kgo.DisableAutoCommit(),
+			kgo.OnPartitionsRevoked(c.lost),
+			kgo.OnPartitionsLost(c.lost),
+		}...)
 	}
 
 	cl, err := kgo.NewClient(opts...)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create kafka client: %w", err)
 	}
 
-	return &FranzConsumer{
-		client: cl,
-		acker:  newBatchAcker(cl, 1000),
-		iter:   &kgo.FetchesRecordIter{}, // empty iterator is done
-	}, nil
+	c.client = cl
+	c.acker = newBatchAcker(cl, 1000)
+
+	return c, nil
+}
+
+func (c *FranzConsumer) lost(ctx context.Context, cl *kgo.Client, lost map[string][]int32) {
+	// TODO: put a proper retry backoff loop here.
+	for c.acker.curBatchIndex != 0 {
+		sdk.Logger(ctx).Warn().Msgf("partitions revoked or lost, waiting until current batch of records are acked and committed...")
+		time.Sleep(50 * time.Millisecond)
+	}
```

Comment:
One concern is that we want to make sure that the current batch of records were acked, so we can go ahead and commit them, then continue processing - this currently loops infinitely, as some records aren't acked. Need to continue debugging.


In general, this is aimed to help prevent duplicates. But if we are okay with at-least-once processing, then, we don't need to implement `OnPartitionsRevoked` / `OnPartitionsLost`: https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#Client.CommitRecords